junit-platform-engine/src/main/java/org/junit/platform/engine/support/store/Namespace.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2015-2025 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.engine.support.store;
+
+import static java.util.Collections.unmodifiableList;
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+import static org.apiguardian.api.API.Status.STABLE;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.util.Preconditions;
+
+@API(status = EXPERIMENTAL, since = "5.13")
+public class Namespace {
+
+	/**
+	 * The default, global namespace which allows access to stored data from
+	 * all extensions.
+	 */
+	public static final Namespace GLOBAL = Namespace.create(new Object());
+
+	/**
+	 * Create a namespace which restricts access to data to all extensions
+	 * which use the same sequence of {@code parts} for creating a namespace.
+	 *
+	 * <p>The order of the {@code parts} is significant.
+	 *
+	 * <p>Internally the {@code parts} are compared using {@link Object#equals(Object)}.
+	 */
+	public static Namespace create(Object... parts) {
+		Preconditions.notEmpty(parts, "parts array must not be null or empty");
+		Preconditions.containsNoNullElements(parts, "individual parts must not be null");
+		return create(Arrays.asList(parts));
+	}
+
+	/**
+	 * Create a namespace which restricts access to data to all extensions
+	 * which use the same sequence of {@code objects} for creating a namespace.
+	 *
+	 * <p>The order of the {@code objects} is significant.
+	 *
+	 * <p>Internally the {@code objects} are compared using {@link Object#equals(Object)}.
+	 */
+	public static Namespace create(List<Object> objects) {
+		Preconditions.notEmpty(objects, "objects list must not be null or empty");
+		Preconditions.containsNoNullElements(objects, "individual objects must not be null");
+		return new Namespace(unmodifiableList(objects));
```

Comment:
We need to create a copy here to avoid being affected by any potential change to the original list.

```suggestion
		return new Namespace(new ArrayList<>(objects));
```

Reply:
Got it!

I’ll make the same changes to the other classes as well.